File: 14 Longest Common Prefix.js
/***********************************************************************************
 * 14 Longest Common Prefix
 * Write a function to find the longest common prefix string amongst
 * an array of strings.
 * Alan Liang, icescut1986@hotmail.com
 * 以第一个数组元素为基础判断剩下的元素在相同位置上的字符是否一致
 * 复杂度O(n)
***********************************************************************************/
/**
 * @param {string[]} strs
 * @return {string}
 */
var longestCommonPrefix = function(strs) {
    var len = strs.length;
    if(len === 0) return "";
    var r = [];
    //遍历第一个字符串元素的所有字符
    for(var i = 0, l1 = strs[0].length; i < l1; ++i) {
        //遍历第二个到最后一个字符串的相同位置
        for(var j = 1; j < len; ++j) {
            if(strs[j][i] !== strs[0][i]) return strs[0].substr(0, i);
        }
    }
    return strs[0];
};

File: 14-Longest-Common-Prefix.cpp
/***********************************************************************************
 * 14 Longest Common Prefix
 * Write a function to find the longest common prefix string amongst
 * an array of strings.
 * Alan Liang, icescut1986@hotmail.com
***********************************************************************************/
#include <iostream>
#include <vector>
using namespace std;

class Solution
{
public:
    string longestCommonPrefix(vector<string> &strs)
    {
        if (strs.empty()) return "";
        int i, j;
        //string r = "";
        for (i = 0; i < strs[0].size(); ++i)
        {
            for (j = 1; j < strs.size(); ++j)
            {
                //no need to compare \0
                if (strs[j][i] != strs[0][i]) return strs[0].substr(0, i);
            }
        }
        return strs[0].substr(0, i);
    }
};

int main()
{
    vector<string> strs =
    {
        "hello",
        "hill",
        "hot"
    };

    vector<string> strs2 =
    {
        "hello",
        "hello",
        "hello"
    };

    vector<string> strs3 =
    {
        "hello",
        "   ",
        "hello"
    };

    vector<string> strs4 =
    {
        "hello",
        "",
        "hello"
    };

    vector<string> strs5 =
    {
        ""
    };

    Solution s;
    cout << s.longestCommonPrefix(strs) << endl;
    cout << s.longestCommonPrefix(strs2) << endl;
    cout << s.longestCommonPrefix(strs3) << endl;
    cout << s.longestCommonPrefix(strs4) << endl;
    cout << s.longestCommonPrefix(strs5) << endl;
    return 0;
}
